File: 0501-0600/0567-Permutation in String/0567-Permutation in String.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s2.size() < s1.size()) {
            return false;
        }
        vector<int> table(26);
        int count = 0;
        for (char c : s1) {
            if (++table[c - 'a'] == 1) {
                ++count;
            }
        }
        
        int start = 0;
        for (int i = 0; i < s2.size(); ++i) {
            if (--table[s2[i] - 'a'] == 0) {
                --count;
                if (count == 0 && i - start + 1 == s1.size()) {
                    return true;
                }
                
                while (count == 0) {
                    if (++table[s2[start] - 'a'] == 1) {
                        count += 1;
                    }
                    else if (i - start == s1.size()) {
                        return true;
                    }
                    ++start;
                }
            }
        }   
        
        return false;
    }
};


class Solution2 {
public:
    bool checkInclusion(string s1, string s2) {
        if (s2.size() < s1.size()) {
            return false;
        }
        vector<int> table(26);
        int count = s1.size();
        for (char c : s1) {
            ++table[c - 'a'];
        }
        
        int start = 0;
        for (int i = 0; i < s2.size(); ++i) {
            if (--table[s2[i] - 'a'] >= 0) {
                --count;
                if (count == 0) {
                    return true;
                }
            }
            
            if (i - start + 1 == s1.size()) {
                if (++table[s2[start++] - 'a'] > 0) {
                    ++count;
                }
            }
        }   
        
        return false;
    }
};

File: 0501-0600/0567-Permutation in String/0567-Permutation in String.py
import collections
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s2) < len(s1):
            return False
        table = collections.Counter(s1)
        count = len(table)
        start = 0
        for i in range(len(s2)):
            table[s2[i]] -= 1
            if table[s2[i]] == 0:
                count -= 1
                if count == 0 and i - start + 1 == len(s1):
                    return True
                    
                while count == 0:
                    table[s2[start]] += 1
                    if table[s2[start]] == 1:
                        count += 1
                    elif i - start == len(s1):
                        return True
                    start += 1
        
        return False


class Solution2:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s2) < len(s1):
            return False
        table = collections.Counter(s1)
        count = len(s1)
        start = 0
        for i in range(len(s2)):
            table[s2[i]] -= 1
            if table[s2[i]] >= 0:
                count -= 1
                if count == 0:
                    return True
                
            if i - start + 1 == len(s1):
                table[s2[start]] += 1
                if table[s2[start]] > 0:
                    count += 1
                start += 1
        
        return False
